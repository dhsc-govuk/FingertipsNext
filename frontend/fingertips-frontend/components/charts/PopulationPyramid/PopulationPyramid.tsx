'use client';

import { HealthDataForArea } from '@/generated-sources/ft-api-client';
import { PopulationPyramidChart } from '@/components/charts/PopulationPyramid/PopulationPyramidChart/PopulationPyramidChart';
import {
  createPyramidPopulationDataFrom,
  PopulationDataForArea,
} from '@/lib/chartHelpers/preparePopulationData';
import { TabContainer } from '@/components/layouts/tabContainer';
import { H3 } from 'govuk-react';
import {
  determineHealthDataForArea,
  seriesDataWithoutGroup,
} from '@/lib/chartHelpers/chartHelpers';
import { SearchParams } from '@/lib/searchStateManager';
import { ArrowExpander } from '@/components/molecules/ArrowExpander';
import { PopulationPyramidChartTable } from './PopulationPyramidTable/PopulationPyramidTableGroup';
import { ChartSelectArea } from '@/components/molecules/ChartSelectArea';
import { areaCodeForEngland } from '@/lib/chartHelpers/constants';
import { StyleChartWrapper } from '@/components/styles/viewPlotStyles/styleChartWrapper';
import {
  determineHeaderTitle,
  determinePopulationDataForArea,
  determineYear,
} from './helpers/populationPyramidHelpers';
import { AreaWithoutAreaType } from '@/lib/common-types';
import { useSearchStateParams } from '@/components/hooks/useSearchStateParams';
import {
  chartTitleConfig,
  ChartTitleKeysEnum,
} from '@/lib/ChartTitles/chartTitleEnums';

interface PyramidPopulationChartViewProps {
  healthDataForAreas: HealthDataForArea[];
  xAxisTitle: string;
  yAxisTitle: string;
  indicatorId: string;
  indicatorName: string;
}
export const PopulationPyramid = ({
  healthDataForAreas,
  xAxisTitle,
  yAxisTitle,
  indicatorId,
  indicatorName,
}: Readonly<PyramidPopulationChartViewProps>) => {
  const searchState = useSearchStateParams();
  const {
    [SearchParams.PopulationAreaSelected]: populationAreaSelected,
    [SearchParams.GroupSelected]: groupSelected,
    [SearchParams.AreaTypeSelected]: areaTypeSelected,
  } = searchState;

  const healthdataWithoutGroup = seriesDataWithoutGroup(
    healthDataForAreas,
    groupSelected,
    true
  );

  const healthDataForAreaSelected = determineHealthDataForArea(
    healthdataWithoutGroup,
    populationAreaSelected
  );

  if (!healthDataForAreaSelected?.healthData.length) return null;

  const { areas, benchmark, group } = createPyramidPopulationDataFrom(
    healthDataForAreas,
    groupSelected ?? ''
  );

  const availableAreas: AreaWithoutAreaType[] = healthdataWithoutGroup
    .map((area) => {
      if (area.areaCode && area.areaName) {
        return {
          code: area.areaCode,
          name: area.areaName,
        };
      }
    })
    .filter((area) => area !== undefined);

  const areasForPopulationData: PopulationDataForArea[] = [];
  areasForPopulationData.push(...areas);
  if (benchmark) areasForPopulationData.push(benchmark);
  if (group) areasForPopulationData.push(group);

  const populationDataForSelectedArea = determinePopulationDataForArea(
    areasForPopulationData,
    availableAreas,
    populationAreaSelected
  );

  if (!populationDataForSelectedArea) return null;

  const benchmarkToUse =
    populationDataForSelectedArea.areaCode !== areaCodeForEngland
      ? benchmark
      : undefined;

  const groupToUse =
    populationDataForSelectedArea.areaCode !== areaCodeForEngland
      ? group
      : undefined;

  const period = determineYear(healthDataForAreaSelected?.healthData ?? []);

  if (!period) return null;

  const title = determineHeaderTitle(
    healthDataForAreaSelected,
    areaTypeSelected,
    period
  );

  return (
    <div data-testid="populationPyramidWithTable-component">
      <StyleChartWrapper>
        <H3
          style={{ fontSize: '24px' }}
          id={ChartTitleKeysEnum.PopulationPyramid}
        >
          {chartTitleConfig[ChartTitleKeysEnum.PopulationPyramid].title}
        </H3>
        <ArrowExpander
          openTitle="Show population data"
          closeTitle="Hide population data"
        >
          <div>
            <ChartSelectArea
              availableAreas={availableAreas}
              chartAreaSelectedKey={SearchParams.PopulationAreaSelected}
            />
            <TabContainer
              id="pyramidChartAndTableView"
              items={[
                {
                  id: 'populationPyramidChart',
                  title: 'Population pyramid',
                  content: (
                    <PopulationPyramidChart
                      title={title}
                      dataForSelectedArea={populationDataForSelectedArea}
                      dataForGroup={groupToUse}
                      dataForBenchmark={benchmarkToUse}
                      xAxisTitle={xAxisTitle}
                      yAxisTitle={yAxisTitle}
                    />
                  ),
                },
                {
                  id: 'populationPyramidTable',
                  title: 'Table',
                  content: (
                    <PopulationPyramidChartTable
                      title={title}
                      healthDataForArea={populationDataForSelectedArea}
                      benchmarkData={benchmarkToUse}
                      groupData={groupToUse}
                      indicatorId={indicatorId}
                      indicatorName={indicatorName}
                      period={period}
                    />
                  ),
                },
              ]}
            />
          </div>
        </ArrowExpander>
      </StyleChartWrapper>
    </div>
  );
};
