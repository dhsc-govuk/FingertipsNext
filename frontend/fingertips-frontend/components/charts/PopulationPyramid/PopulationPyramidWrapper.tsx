'use client';

import { PopulationPyramid } from '@/components/charts/PopulationPyramid/PopulationPyramid';
import { usePopulationPyramidData } from '@/components/charts/PopulationPyramid/hooks/usePopulationPyramidData';

export function PopulationPyramidWrapper() {
  const { healthData } = usePopulationPyramidData();
  if (!healthData) return null;

  return (
    <PopulationPyramid
      healthDataForAreas={healthData.areaHealthData ?? []}
      xAxisTitle="Age"
      yAxisTitle="Percentage of total population"
      indicatorId={String(healthData.indicatorId ?? 0)}
      indicatorName={healthData.name ?? ''}
    />
  );
}
