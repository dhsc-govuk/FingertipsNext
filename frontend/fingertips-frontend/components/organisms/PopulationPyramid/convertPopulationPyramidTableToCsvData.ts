import { CsvHeader } from '@/components/molecules/Export/export.types';
import { PopulationDataForArea } from '@/lib/chartHelpers/preparePopulationData';
import { CsvData, CsvRow } from '@/lib/downloadHelpers/convertToCsv';

export function convertPopulationPyramidTableToCsvData(
  period: number,
  populationDataForArea: PopulationDataForArea,
  indicatorId?: string,
  indicatorName?: string,
  populationDataForBenchmark?: PopulationDataForArea,
  populationDataForGroup?: PopulationDataForArea
): CsvData {
  if (!indicatorId || !indicatorName) {
    throw new Error('IndicatorID and IndicatorName are required');
  }
  const header: CsvRow = [
    CsvHeader.IndicatorId,
    CsvHeader.IndicatorName,
    CsvHeader.Period,
    CsvHeader.Area,
    CsvHeader.AreaCode,
    CsvHeader.AgeRange,
    CsvHeader.Male,
    CsvHeader.Female,
  ];

  const rows: CsvData = [];
  if (populationDataForBenchmark)
    rows.push(
      ...convertPopulationPyramidTableAreaToCsvRow(
        indicatorId,
        indicatorName,
        period,
        populationDataForBenchmark
      )
    );
  if (populationDataForGroup)
    rows.push(
      ...convertPopulationPyramidTableAreaToCsvRow(
        indicatorId,
        indicatorName,
        period,
        populationDataForGroup
      )
    );
  rows.push(
    ...convertPopulationPyramidTableAreaToCsvRow(
      indicatorId,
      indicatorName,
      period,
      populationDataForArea
    )
  );

  return [header, ...rows];
}
function convertPopulationPyramidTableAreaToCsvRow(
  indicatorId: string,
  indicatorName: string,
  period: number,
  populationDataForArea: PopulationDataForArea
): CsvData {
  return populationDataForArea.ageCategories.map((ageCategory, i) => [
    indicatorId,
    indicatorName,
    period,
    populationDataForArea.areaName,
    populationDataForArea.areaCode,
    ageCategory,
    populationDataForArea.maleSeries[i],
    populationDataForArea.femaleSeries[i],
  ]);
}
