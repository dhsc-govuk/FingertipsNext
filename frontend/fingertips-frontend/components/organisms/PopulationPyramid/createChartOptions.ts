import {
  computeDataPercentages,
  PopulationDataForArea,
} from '@/lib/chartHelpers/preparePopulationData';
import Highcharts, { SeriesOptionsType } from 'highcharts';
import { pointFormatterHelper } from '@/lib/chartHelpers/pointFormatterHelper';
import { generatePopPyramidTooltipForPoint } from '.';
import { GovukColours } from '@/lib/styleHelpers/colours';

const createChartSeriesOptions = (
  xAxisTitle: string,
  yAxisTitle: string,
  dataForArea: PopulationDataForArea,
  accessibilityLabel: string
): Highcharts.Options => {
  const femaleSeries = computeDataPercentages(
    dataForArea.femaleSeries,
    dataForArea.total
  );
  const maleSeries = computeDataPercentages(
    dataForArea.maleSeries,
    dataForArea.total
  );
  const maxTick = Math.abs(Math.max(...femaleSeries, ...maleSeries));
  return {
    chart: {
      type: 'bar',
      height: 800,
      animation: false,
    },
    plotOptions: {
      series: {
        stacking: 'normal',
        crisp: false,
        animation: false,
      },
    },
    credits: { enabled: false },
    title: { style: { display: 'none' } },
    legend: {
      verticalAlign: 'top',
      layout: 'horizontal',
      alignColumns: true,
      reversed: true,
      itemStyle: {
        fontSize: '19px',
      },
    },
    xAxis: [
      {
        categories: dataForArea.ageCategories,
        title: {
          text: xAxisTitle,
          align: 'high',
          offset: 2,
          rotation: 0,
        },
        lineColor: GovukColours.DarkSlateGray,
        tickWidth: 1,
        tickLength: 10,
        tickmarkPlacement: 'on',
        tickColor: GovukColours.DarkSlateGray,
        labels: {
          style: {
            fontSize: 16,
          },
        },
      },
      {
        opposite: true,
        reversed: false,
        categories: dataForArea.ageCategories,
        linkedTo: 0,
        title: {
          text: xAxisTitle,
          align: 'high',
          offset: 4,
          rotation: 0,
        },
        lineColor: GovukColours.DarkSlateGray,
        tickWidth: 1,
        tickLength: 10,
        tickmarkPlacement: 'on',
        tickColor: GovukColours.DarkSlateGray,
        accessibility: {
          description: '{xAxisTitle} degrees {series.name}',
        },
        labels: {
          style: {
            fontSize: 16,
          },
        },
      },
    ],
    yAxis: {
      title: {
        text: yAxisTitle,
      },
      min: -maxTick,
      max: maxTick,
      lineWidth: 1,
      lineColor: GovukColours.DarkSlateGray,
      tickWidth: 1,
      tickLength: 5,
      tickColor: GovukColours.DarkSlateGray,
      gridLineWidth: 0,
      labels: {
        format: '{abs value}%',
        align: 'center',
        style: {
          fontSize: 16,
        },
      },
      accessibility: {
        enabled: false,
        description: accessibilityLabel,
      },
      tickInterval: 1,
    },
    tooltip: {
      padding: 10,
      headerFormat: `<div style="margin:0px; padding:0px;">
            <span style="font-weight: bold; display: block;">
            ${dataForArea.areaName}
            </span>
        <span>Age {key}</span><div>`,
      pointFormatter: function (this: Highcharts.Point) {
        return pointFormatterHelper(this, generatePopPyramidTooltipForPoint);
      },
      useHTML: true,
    },
    series: [
      {
        name: 'Female',
        type: 'bar',
        data: femaleSeries,
        xAxis: 0,
        color: GovukColours.Female,
        pointWidth: 17,
        dataLabels: {
          enabled: false,
          inside: false,
          format: '{(abs point.y):.1f}%',
          color: GovukColours.Black,
          style: {
            fontWeight: 'light',
            fontSize: '16px',
          },
        },
      },
      {
        name: 'Male',
        type: 'bar',
        data: maleSeries.map((datapoint) => -datapoint),
        xAxis: 1,
        color: GovukColours.Male,
        pointWidth: 17,
        dataLabels: {
          enabled: false,
          inside: false,
          format: '{(abs point.y):.1f}%',
          color: GovukColours.Black,
          style: {
            fontWeight: 'light',
            fontSize: '16px',
          },
        },
      },
    ],
    accessibility: {
      enabled: false,
      description: accessibilityLabel,
    },
  };
};

const createAdditionalChartSeries = (
  dataForBenchmark: PopulationDataForArea | undefined,
  dataForGroup: PopulationDataForArea | undefined
): Array<SeriesOptionsType> => {
  const series: Array<SeriesOptionsType> = [];

  if (dataForGroup) {
    const femaleGroupSeries = computeDataPercentages(
      dataForGroup.femaleSeries,
      dataForGroup.total
    );
    const maleGroupSeries = computeDataPercentages(
      dataForGroup.maleSeries,
      dataForGroup.total
    );

    series.push(
      {
        name: `Group: ${dataForGroup.areaName}`,
        type: 'line',
        data: femaleGroupSeries,
        stack: 2,
        color: GovukColours.Turquoise,
        dashStyle: 'Dash',
        marker: { symbol: 'diamond' },
        dataLabels: { enabled: false },
      },
      {
        name: `Group: ${dataForGroup.areaName}`,
        type: 'line',
        stack: 4,
        data: maleGroupSeries.map((datapoint) => -datapoint),
        color: GovukColours.Turquoise,
        dashStyle: 'Dash',
        marker: { symbol: 'diamond' },
        dataLabels: { enabled: false },
        showInLegend: false,
      }
    );
  }

  if (dataForBenchmark) {
    const femaleBenchmarkSeries = computeDataPercentages(
      dataForBenchmark.femaleSeries,
      dataForBenchmark.total
    );
    const maleBenchmarkSeries = computeDataPercentages(
      dataForBenchmark.maleSeries,
      dataForBenchmark.total
    );
    series.push(
      {
        name: `Benchmark: ${dataForBenchmark.areaName}`,
        data: femaleBenchmarkSeries,
        type: 'line',
        stack: 1,
        color: GovukColours.CharcoalGray,
        dashStyle: 'Solid',
        marker: { symbol: 'circle' },
      },
      {
        name: `Benchmark: ${dataForBenchmark.areaName}`,
        data: maleBenchmarkSeries.map((datapoint) => -datapoint),
        type: 'line',
        stack: 3,
        color: GovukColours.CharcoalGray,
        dashStyle: 'Solid',
        marker: { symbol: 'circle' },
        showInLegend: false,
      }
    );
  }
  return series;
};

export const createChartPyramidOptions = (
  xAxisTitle: string,
  yAxisTitle: string,
  accessibilityLabel: string,
  dataForSelectedArea: PopulationDataForArea,
  dataForBenchmark?: PopulationDataForArea,
  dataForSelectedGroup?: PopulationDataForArea
): Highcharts.Options => {
  const populationPyramidOptions: Highcharts.Options = createChartSeriesOptions(
    xAxisTitle,
    yAxisTitle,
    dataForSelectedArea,
    accessibilityLabel
  );
  if (!populationPyramidOptions.series) {
    return populationPyramidOptions;
  }
  if (populationPyramidOptions.series.length > 0) {
    const seriesOptions = createAdditionalChartSeries(
      dataForBenchmark,
      dataForSelectedGroup
    );
    if (seriesOptions)
      seriesOptions.forEach((series) => {
        populationPyramidOptions.series?.push(series);
      });
  }
  return populationPyramidOptions;
};
