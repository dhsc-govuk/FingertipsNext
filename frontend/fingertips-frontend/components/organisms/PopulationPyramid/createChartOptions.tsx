import { PopulationDataForArea } from '@/lib/chartHelpers/preparePopulationData';
import Highcharts, { SeriesOptionsType } from 'highcharts';
import { pointFormatterHelper } from '@/lib/chartHelpers/pointFormatterHelper';
import { PopulationData } from '@/lib/chartHelpers/preparePopulationData';

export const createChartSeriesOptions = (
  xAxisTitle: string,
  yAxisTitle: string,
  dataForArea: PopulationDataForArea,
  accessibilityLabel: string
): Highcharts.Options => {
  return {
    chart: {
      type: 'bar',
      height: 800,
    },
    plotOptions: {
      series: {
        stacking: 'normal',
        crisp: false,
      },
    },
    credits: { enabled: false },
    title: { style: { display: 'none' } },
    legend: { verticalAlign: 'top', layout: 'horizontal' },
    xAxis: [
      {
        categories: dataForArea.ageCategories,
        title: {
          text: xAxisTitle,
          align: 'high',
          offset: 2,
          rotation: 0,
        },
        lineColor: '#D7D7D7',
        tickWidth: 1,
        tickLength: 10,
        tickmarkPlacement: 'on',
        tickColor: '#D7D7D7',
      },
      {
        // mirror axis on right side
        opposite: true,
        reversed: false,
        categories: dataForArea.ageCategories,
        linkedTo: 0,
        title: {
          text: xAxisTitle,
          align: 'high',
          offset: 4,
          rotation: 0,
        },
        lineColor: '#D7D7D7',
        tickWidth: 1,
        tickLength: 10,
        tickmarkPlacement: 'on',
        tickColor: '#D7D7D7',
        accessibility: {
          description: '{xAxisTitle} degrees {series.name}',
        },
      },
    ],
    yAxis: {
      title: {
        text: yAxisTitle,
      },
      lineWidth: 1,
      lineColor: '#D7D7D7',
      tickWidth: 1,
      tickLength: 5,
      tickColor: '#D7D7D7',
      gridLineWidth: 0,
      labels: {
        format: '{abs value} %',
        align: 'center',
      },
      accessibility: {
        enabled: false,
        description: accessibilityLabel,
      },
      tickInterval: 1,
      // tickPositioner: (axis: Axis) => {
      //   if (!axis.isXAxis) {
      //     return [
      //       new Tick(axis, -100, "bar", false),
      //       new Tick(axis, 0, "bar", false),
      //       new Tick(axis, 100, "bar", false)
      //     ]

      //   }
      // },
    },
    tooltip: {
      padding: 10,
      headerFormat: `<div style="margin:0px; padding:0px;">
            <span style="font-weight: bold; display: block;">
            ${dataForArea.areaName}
            </span>
        <span display:inline-block;>Age {key}</span><div>`,
      pointFormatter: function (this: Highcharts.Point) {
        return pointFormatterHelper(this);
      },
      useHTML: true,
    },
    series: [
      {
        name: 'Female',
        type: 'bar',
        data: dataForArea.femaleSeries,
        xAxis: 0,
        color: '#5352BE',
        pointWidth: 17,
      },
      {
        name: 'Male',
        type: 'bar',
        data: dataForArea.maleSeries.map((datapoint) => -datapoint),
        xAxis: 1,
        color: '#57AEF8',
        pointWidth: 17,
      },
    ],
    accessibility: {
      enabled: false,
      description: accessibilityLabel,
    },
  };
};

export const createAdditionalChartSeries = (
  healthIndicatorData: PopulationData | undefined
): Array<SeriesOptionsType> => {
  const series: Array<SeriesOptionsType> = [];
  if (!healthIndicatorData) {
    return [];
  }
  if (healthIndicatorData.dataForEngland) {
    series.push(
      {
        name: 'England',
        data: healthIndicatorData.dataForEngland.femaleSeries,
        type: 'line',
        color: '#3D3D3D',
        dashStyle: 'Solid',
        marker: { symbol: 'circle' },
      },
      {
        name: 'England',
        data: healthIndicatorData.dataForEngland.maleSeries.map(
          (datapoint) => -datapoint
        ),
        type: 'line',
        color: '#3D3D3D',
        dashStyle: 'Solid',
        marker: { symbol: 'circle' },
        showInLegend: false,
      }
    );
  }
  if (healthIndicatorData.dataForBaseline) {
    series.push(
      {
        name: 'Baseline',
        type: 'line',
        data: healthIndicatorData.dataForBaseline.femaleSeries,
        color: '#28A197',
        dashStyle: 'Dash',
        marker: { symbol: 'diamond' },
        dataLabels: { enabled: false },
      },
      {
        name: 'Baseline',
        type: 'line',
        data: healthIndicatorData.dataForBaseline.maleSeries.map(
          (datapoint) => -datapoint
        ),
        color: '#28A197',
        dashStyle: 'Dash',
        marker: { symbol: 'diamond' },
        dataLabels: { enabled: false },
        showInLegend: false,
      }
    );
  }
  return series;
};
