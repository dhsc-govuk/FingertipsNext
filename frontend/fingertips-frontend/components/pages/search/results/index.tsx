'use client';

import { BackLink, H1, Paragraph } from 'govuk-react';

export function SearchResults({ indicator }: Readonly<{ indicator: string }>) {
  return (
    <>
      <BackLink href={`/search?indicator=${indicator}`} />
      <H1>Search results</H1>
      <Paragraph>{`You searched for indicator "**${indicator}**"`}</Paragraph>
    </>
  );
}
