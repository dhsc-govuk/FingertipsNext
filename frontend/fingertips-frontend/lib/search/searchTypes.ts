/*********************************************************************************
 * This file includes duplicate definitions from the search-setup/src/constants.ts
 * These should be updated to take definitions from there rather than duplicating them
 */

export const INDICATOR_SEARCH_SCORING_PROFILE = 'basicScoringProfile';
export const INDICATOR_SEARCH_INDEX_NAME = 'indicator-search-index';
export const AREA_SEARCH_INDEX_NAME = 'area-search-index';
export const AREA_SEARCH_SUGGESTER_NAME = 'areaSuggester';

export type IndicatorDocument = {
  indicatorId: string;
  name: string;
  definition: string;
  dataSource: string;
  latestDataPeriod: string;
  lastUpdated: Date;
};

export type AreaDocument = {
  areaCode: string;
  areaType: string;
  areaName: string;
};

export interface IIndicatorSearchService {
  searchWith(searchTerm: string): Promise<IndicatorDocument[]>;
}

export interface IAreaSearchService {
  getAreaSuggestions(partialAreaName: string): Promise<AreaDocument[]>;
}
