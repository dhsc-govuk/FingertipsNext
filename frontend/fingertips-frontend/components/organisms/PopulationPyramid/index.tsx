'use client';

import Highcharts from 'highcharts';
import { PopulationDataForArea } from '@/lib/chartHelpers/preparePopulationData';
import { createChartPyramidOptions } from './createChartOptions';
import { HighChartsWrapper } from '@/components/molecules/HighChartsWrapper/HighChartsWrapper';
import { ExportOptionsButton } from '@/components/molecules/Export/ExportOptionsButton';
import { ExportOnlyWrapper } from '@/components/molecules/Export/ExportOnlyWrapper';
import { ExportCopyright } from '@/components/molecules/Export/ExportCopyright';
import { ChartTitle } from '@/components/atoms/ChartTitle/ChartTitle';

interface PyramidChartProps {
  title: string;
  dataForSelectedArea: PopulationDataForArea;
  dataForBenchmark?: PopulationDataForArea;
  dataForGroup?: PopulationDataForArea;
  xAxisTitle: string;
  yAxisTitle: string;
  accessibilityLabel?: string;
}

export function PopulationPyramid({
  title,
  dataForSelectedArea,
  dataForBenchmark,
  dataForGroup,
  xAxisTitle,
  yAxisTitle,
  accessibilityLabel,
}: Readonly<PyramidChartProps>) {
  Highcharts.Templating.helpers.abs = (value) => Math.abs(value);

  const populationPyramidOptions: Highcharts.Options =
    createChartPyramidOptions(
      xAxisTitle,
      yAxisTitle,
      accessibilityLabel ?? '',
      dataForSelectedArea,
      dataForBenchmark,
      dataForGroup
    );
  const id = 'populationPyramidChart';
  return (
    <>
      <div id={id} data-testid={`${id}-component`}>
        <ChartTitle>{title}</ChartTitle>
        <HighChartsWrapper
          testId={'highcharts-react-component-populationPyramid'}
          chartOptions={populationPyramidOptions}
        />
        <ExportOnlyWrapper>
          <ExportCopyright />
        </ExportOnlyWrapper>
      </div>
      <ExportOptionsButton
        targetId={id}
        chartOptions={populationPyramidOptions}
      />
    </>
  );
}
