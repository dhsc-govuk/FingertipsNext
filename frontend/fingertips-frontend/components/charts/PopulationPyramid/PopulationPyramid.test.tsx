// MUST BE AT THE TOP DUE TO HOISTING OF MOCKED MODULES
import { mockSetIsLoading } from '@/mock/utils/mockUseLoadingState';
import { mockUseSearchStateParams } from '@/mock/utils/mockUseSearchStateParams';
import { mockUsePathname } from '@/mock/utils/mockNextNavigation';
//
import { render, screen, fireEvent, within } from '@testing-library/react';
import { PeriodType, ReportingPeriod } from '@/generated-sources/ft-api-client';
import { AreaDocument } from '@/lib/search/searchTypes';
import { SearchStateParams } from '@/lib/searchStateManager';
import { PopulationPyramid } from '@/components/charts/PopulationPyramid/PopulationPyramid';
import {
  chartTitleConfig,
  ChartTitleKeysEnum,
} from '@/lib/ChartTitles/chartTitleEnums';
import { mockHealthDataForArea } from '@/mock/data/mockHealthDataForArea';
import { mockIndicatorSegment } from '@/mock/data/mockIndicatorSegment';
import { mockHealthDataPoint } from '@/mock/data/mockHealthDataPoint';
import { mockDatePeriod } from '@/mock/data/mockDatePeriod';

const mockSearchState: SearchStateParams = {};

mockUsePathname.mockReturnValue('some-path-name');
mockSetIsLoading.mockReturnValue(false);
mockUseSearchStateParams.mockReturnValue(mockSearchState);

// Mock dependencies
vi.mock(
  '@/components/charts/PopulationPyramid/PopulationPyramidChart/PopulationPyramidChart',
  () => ({
    PopulationPyramidChart: () => <div data-testid="population-pyramid"></div>,
  })
);
vi.mock('@/components/molecules/SelectInputField', () => ({
  AreaSelectInputField: ({
    onSelected,
  }: {
    onSelected: (area: Omit<AreaDocument, 'areaType'>) => void;
  }) => (
    <button
      data-testid="select-input"
      onClick={() => onSelected({ areaCode: '123', areaName: 'Test Area' })}
    >
      Select Area
    </button>
  ),
}));

const localMockHealthDataForArea = [
  mockHealthDataForArea({
    areaCode: '123',
    areaName: 'Test Area 123',
  }),
  mockHealthDataForArea({
    areaCode: '124',
    areaName: 'Test Area 124',
  }),
];

function testRender(dataForArea = localMockHealthDataForArea) {
  return render(
    <PopulationPyramid
      healthDataForAreas={dataForArea}
      xAxisTitle="Age"
      yAxisTitle="Percentage of population"
      indicatorId={'1'}
      indicatorName={'Indicator'}
    />
  );
}

describe('PopulationPyramidWithTable', () => {
  test('renders component with default title', () => {
    testRender();
    expect(
      screen.getByText(
        chartTitleConfig[ChartTitleKeysEnum.PopulationPyramid].title
      )
    ).toBeInTheDocument();
  });

  test('renders tabs correctly', () => {
    testRender();
    expect(screen.getByText('Show population data')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Show population data'));
    expect(screen.getByText('Hide population data')).toBeInTheDocument();
  });

  it('test that we can clicked the expander', async () => {
    testRender();
    const populationPyramid = screen.getByTestId(
      'populationPyramidWithTable-component'
    );
    const expander = await within(populationPyramid).findByText(
      'Show population data'
    );
    fireEvent.click(expander);
    expect(screen.getByText('Hide population data')).toBeInTheDocument();
  });

  it('should render the availableAreas as options in the Select an area dropdown', async () => {
    testRender();

    fireEvent.click(screen.getByText('Show population data'));

    const populationAreasDropDown = screen.getByRole('combobox', {
      name: 'Select an area',
    });
    const populationAreasDropDownOptions = within(
      populationAreasDropDown
    ).getAllByRole('option');

    expect(populationAreasDropDown).toBeInTheDocument();
    expect(populationAreasDropDownOptions).toHaveLength(2);
    populationAreasDropDownOptions.forEach((option, i) => {
      expect(option.textContent).toBe(localMockHealthDataForArea[i].areaName);
    });
  });

  it('should not render if health data is empty', () => {
    testRender([]);
    expect(
      screen.queryByTestId('populationPyramidWithTable-component')
    ).not.toBeInTheDocument();
  });

  it('should show No "data" message if no data for the area', () => {
    const mockHealthDataForWithNoDatafForArea = [
      mockHealthDataForArea({
        indicatorSegments: [
          mockIndicatorSegment({
            healthData: [],
          }),
        ],
      }),
    ];

    testRender(mockHealthDataForWithNoDatafForArea);
    fireEvent.click(screen.getByText('Show population data'));

    expect(
      screen.queryByTestId('populationPyramidWithTable-component')
    ).toBeInTheDocument();
    expect(
      screen.getByText('No population data for this area')
    ).toBeInTheDocument();
  });
});
