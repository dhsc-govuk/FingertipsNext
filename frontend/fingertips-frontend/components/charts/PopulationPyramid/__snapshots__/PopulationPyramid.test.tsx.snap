// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopulationPyramidWithTable > take a snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="populationPyramidWithTable-component"
  >
    <div
      class="sc-jxOSlx imMZIm"
    >
      <h3
        class="heading__StyledHeading-sc-1dsobun-0 bRbIzR"
        id="population-pyramid-chart"
        style="font-size: 24px;"
      >
        Related population data
      </h3>
      <span
        class="sc-fqkvVR ieEIxa"
      >
        <svg
          class="sc-eqUAAy hOozcF"
          data-testid="arrow-toggle-button"
          height="30"
          viewBox="0 0 25 25"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
            stroke="#1D70B8"
            stroke-width="0.5"
          />
          <path
            d="M8 10l4 4 4-4"
            fill="none"
            stroke="#1D70B8"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <label
          class="src__Label-sc-iqzvxn-0 sc-dcJsrY cjKMYe knleWj"
          fill="#1D70B8"
        >
          Show population data
        </label>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
